File: apogee-backend/apogee/utils/cleaner.py
import re

def clean_text(text: str) -> str:
    # Collapse runs of blank lines into a single newline
    text = re.sub(r'\n{3,}', '\n\n', text)
    # Collapse horizontal whitespace (spaces/tabs) but preserve newlines
    text = re.sub(r'[^\S\n]+', ' ', text)
    # Strip leading/trailing whitespace from each line
    text = '\n'.join(line.strip() for line in text.splitlines())
    return text.strip()

File: apogee-backend/src/utils/cleaner.js
export function cleanText(text) {
  // Collapse runs of blank lines into a single newline
  let cleaned = text.replace(/\n{3,}/g, "\n\n");
  // Collapse horizontal whitespace (spaces/tabs) but preserve newlines
  cleaned = cleaned.replace(/[^\S\n]+/g, " ");
  // Strip leading/trailing whitespace from each line
  cleaned = cleaned
    .split("\n")
    .map((line) => line.trim())
    .join("\n");
  return cleaned.trim();
}
